edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -159,22 +167,28 @@ public async Task UpdatePriorities(IList<uint> priorities, Option<Endpoint> newE
                     EndpointExecutorFsm fsm = new EndpointExecutorFsm(this.Endpoint, checkpointer, this.config);
 
                     // Add it to our dictionary
-                    snapshot.Add(priority, fsm);
+                    updatedSnapshot.Add(priority, fsm);
                 }
                 else
                 {
                     // Update the existing FSM with the new endpoint
-                    EndpointExecutorFsm fsm = snapshot[priority];
+                    EndpointExecutorFsm fsm = updatedSnapshot[priority];
                     await newEndpoint.ForEachAsync(e => fsm.RunAsync(Commands.UpdateEndpoint(e)));
                 }
             }
 
-            this.prioritiesToFsms.Value = snapshot;
-            Events.UpdatePrioritiesSuccess(this, snapshot.Keys.ToList());
+            if (this.prioritiesToFsms.CompareAndSet(snapshot, updatedSnapshot.ToImmutableDictionary()))

```

Comment:
@varunpuranik  I think it is more safe to use CompareAndSet and write an error log if happens.